docs/design/specs/Ecma-335-Augments.md

```diff
@@ -948,3 +949,47 @@ Conversions from floating-point numbers to integral values truncate the number t
+- Add a new paragraph before the paragraph on the three special types. “Byref-like types are value types which may contain managed pointers, or pointers onto the VES stack. Byref-like types have the same restrictions as byrefs. Value types which are marked with the System.Runtime.CompilerServices.IsByRefLikeAttribute attribute are considered to be byref-like types.”
+- Replace “Typed references have the same restrictions as byrefs.” With “Typed references are byref-like types and have the same restrictions as normal byref-like types.”
+- Replace “They can be used for local variable and parameter signatures. The use of these types for fields, method return types, the element type of an array, or in boxing is not verifiable (§I.8.8). These two types are referred to as byref-like types.” With “If a function uses these types only for local variable and parameter signatures, then the use of those types is verifiable. Otherwise these types have the same restrictions as normal byref-like types.”
+### I.8.6.1
+- Remove “(A fifth kind, a local signature (see §I.8.6.1.3) is really a version of a location signature.)”
+
+### I.8.6.1.2
+- Move the contents of the byref constraint paragraph from Section I.8.6.1.3 to a bullet point.
+“- The byref constraint states that the content of the corresponding location is a managed pointer. A managed pointer can point to a local variable, parameter, field of a compound type, or element of an array. However, when a call crosses a remoting boundary (see §I.12.5) a conforming implementation can use a copy-in/copy-out mechanism instead of a managed pointer. Thus programs shall not rely on the aliasing behavior of true pointers. A managed pointer cannot point to another managed pointer, but a managed pointer can point to a byref-like type.”
+
+### I.8.6.1.3
+- Rename this section to ByRef-like and Typed Reference Signatures
+- Remove the first two paragraphs.
+- Remove “In addition, there is one special local signature.”
+- Remove “A typed reference is a full signature in itself and cannot be combined with other constraints. In particular, it is not possible to specify a byref whose type is typed reference.”
+- Replace “This type shall only be used for parameters and local variables. It shall not be boxed, nor shall it be used as the type of a field, element of an array, or return value.” With “A typed reference is considered to be a by-reflike structure”
+- Remove “A typed reference is a full signature in itself and cannot be combined with other constraints. In particular, it is not possible to specify a byref whose type is typed reference.” From the third paragraph
```

Comment:
In the spec under "CLS Rule 14: Typed references are not CLS-compliant", the phase "CLS (framework): This type shall not appear in exported members." seems like it should be softened. I'd assume this type may start to appear in other types, no? I could be misreading what this is stating.

Reply:
Just mark those members as not CLS-compliant, like any members whose type is UInt32.

Reply:
It would still be great if we could just remove the concept of `CLSCompliant`.

C# already doesn't correctly warn everywhere it should (things like `modreq` aren't `CLSCompliant` and there are several cases, including `where T : unmanaged`, where C# doesn't force you to add the attribute).

I, personally, view `CLSCompliant` a lot like "integers don't have to be two's complement representation" in C/C++. From an extensibility point it "makes sense" and allows the language/framework to "target more". However, from a practicality standpoint its a feature that basically doesn't get used and causes pain/headaches and various compat issues because half the tooling doesn't consider it.

Reply:
> It would still be great if we could just remove the concept of CLSCompliant.

The idea of `CLSCompliant` doesn't really exist on .NET Core. It's not really been a part of the tooling plan. Agree we should think about overall deprecating the concept if possible.

Reply:
CLS compliance is really only about API design guidelines.

The value of CLS compliance was discussed multiple times during API review meetings. One of the last discussions - https://github.com/dotnet/runtime/issues/44194#issuecomment-887742046. I believe that the prevalent opinion in the room was that there is still some value in it and we are not ready to make the call to deprecate the concept.

I have tried to delete CLS compliant attributes in .NET Core to make it not really exist on. With the current Roslyn behavior, it would be a source breaking change for everybody who marks their assembly as CLS compliant. The compiler warns or errors when CLS compliant assemblies depend on non-CLS compliant assemblies.